### ISSUE TITLE
MNIST example scores 9.95% — call evolveDir directly with targetError 0.001

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-AI-Examples#268]
Please make sure you've created binary training data then called Creature.evolveDir for at least 5 minutes. I would expect a better result if that's what you've done. 

I suspect you've tried to code the evolution yourself.

We should create a creature with the correct input/output , allow it to automatically initialise ( just use the default no need to pass anything) 

Then call evolveDir with the only options of target error of 0.001 and timeout of 5 minutes, I would expect good evolution within that time, if not may need to increase timeout a bit

https://github.com/stSoftwareAU/NEAT-AI-Examples/blob/Develop/mnist_classification/README.md

## Current Understanding

The published `mnist_classification` example reports only **9.95% test accuracy** (essentially random guessing for 10 classes), contradicting the README claim that the champion "hits the 95% accuracy target". The fix is to delete the hand-rolled `runMinimalSeedEvolveDir` chunking wrapper and call `Creature.evolveDir` once with a minimal options bag, using the full 28×28 / 60k-sample dataset and a 10-minute upper-bound timeout. Whatever accuracy the honest run produces is the result — the README will be cut down to a short "what evolveDir produced in 10 min from a minimal seed" report quoting that real number.

### Accepted scope so far
- Seed is `new Creature(784, 10)` — full 28×28 inputs, 10 outputs, no `NeatOptions` overrides, no warm start.
- Train and test on the full MNIST dataset (60k train / 10k test) with the existing Rust scorer pipeline.
- Call `Creature.evolveDir(dataDir, options)` exactly once. Delete `runMinimalSeedEvolveDir` and its `populationSize: 12` / `maxIterations: 200` overrides.
- `evolveDir` options are limited to `targetError: 0.001` and `timeoutMinutes: 10`.
- Per-generation telemetry rides on existing or planned `evolveDir` options in NEAT-AI — no wrapper loop. If a needed option is missing, raise a follow-up issue against `stSoftwareAU/NEAT-AI` (mirroring NEAT-AI #2629 for evolveRL) with a call-back to update this example when the hooks land. Charts ship in a follow-up PR, not this one.
- README is cut to a short report quoting the real measured final accuracy and runtime — no "95% target" framing, no extended narrative.
- Evidence: a single 10-minute run is enough; quote that number as the result.
- Any limitations identified during this work that block telemetry or quality must be raised as `stSoftwareAU/NEAT-AI` issues with a call-back to update the examples here when fixed.

### Open questions
None. All material questions are settled.

### Assumptions
- "Default" creature initialisation means `new Creature(input, output)` with no third argument and no `NeatOptions` overrides anywhere.
- The Rust scorer can score the full 60k training set per generation within the 10-min budget; if not, that is a NEAT-AI/scorer issue, not an examples issue.
- A from-scratch NEAT topology evolution on 784→10 with 60k samples is unlikely to reach 95% in 10 minutes on CPU — published NEAT-style MNIST work typically runs for hours. Expect a low headline number from the first honest run; the user has accepted "results are the results".

### Related open issues
- #262 — adaptive_mutation classification primitives. Different task (parity/multiplexer), but should share the same `evolveDir(dataDir, { targetError, timeoutMinutes })` call shape so the examples stay consistent.
- #263 — adaptive_mutation rewires evolution loop. Same minimal-seed + single-call pattern this issue adopts. Keep the call shape aligned.
- #264 — adaptive_mutation README/charts refresh. Sets the "real measured numbers, no estimates" README convention this issue follows.
- #220, #223 — cart_pole / maze_navigation audits. Establish the per-generation CSV + SVG telemetry convention for interactive environments; MNIST is supervised-batch via `evolveDir`, so it inherits the spirit (real numbers only) but waits on NEAT-AI library hooks rather than hand-rolling per-generation collection.
- #235 — supervised-batch vs event-driven docs. No direct overlap, but this PR is a canonical supervised-batch example.
- `stSoftwareAU/NEAT-AI` — no existing open issue for `evolveDir` per-generation statistics (the analogous evolveRL issue is #2629). A follow-up will be filed there as part of this work, with a call-back to update this example when hooks land.

**Comment by stservice** [comment]
## Grill-Me Round 1

**TL;DR:** Confirming you want the example stripped to a single `evolveDir` call with only `targetError: 0.001` + `timeoutMinutes: 5`, and what counts as a "good" final accuracy.

### Understanding
The current `mnist_classification.ts` wraps `evolveDir` in a hand-rolled chunking loop (`runMinimalSeedEvolveDir`) with `populationSize: 12`, `maxIterations: 200`, and `targetError: 0.02`. The latest run hit the `maxIterations` cap, not the time budget, and ended at 9.95% test accuracy. You want this replaced with a single `evolveDir` call passing only `targetError` and `timeoutMinutes`.

### Questions

1. Strip `evolveDir` options to **only** `{ targetError: 0.001, timeoutMinutes: 5 }`?
   - [x] Yes — remove `populationSize`, `maxIterations`, and the chunking loop entirely
   - [ ] Yes, but keep `populationSize` configurable
   - [ ] Keep the chunking loop for per-generation telemetry
   - [x] other — please describe in a reply

2. What counts as "good" final test accuracy for a PR to be acceptable?
   - [ ] ≥ 50% (clearly beats chance)
   - [ ] ≥ 80%
   - [ ] ≥ 90%
   - [ ] ≥ 95% (matches the MLP baseline)
   - [x] other — please describe in a reply

3. If 5 minutes isn't enough, what's the upper bound for the timeout?
   - [x] 10 minutes
   - [ ] 30 minutes
   - [ ] 60 minutes
   - [ ] No cap — let it run until target error
   - [ ] other — please describe in a reply

4. Keep the current input shape and dataset size?
   - [ ] Keep 14×14 down-sample + 1024 training samples
   - [ ] Keep 14×14, increase to full 60k samples
   - [ ] Use full 28×28 (784 inputs), keep 1024 samples
   - [x] Use full 28×28 + full 60k samples
   - [ ] other — please describe in a reply

Tick the boxes that apply (the GitHub mobile app lets you tap each one), then add any free-form notes in a reply.

**⏳ Awaiting your reply.** The Vibe Coder is waiting for your response on this issue to continue grilling.

**Comment by nleck** [comment]
Q1: remove the "chunking loop" for now. Please examine the NEAT-AI issues and options, we will add the needed options to gather the required statistics. If the existing ( or planned ) options in NEAT-AI will not give you want please raise issues in NEAT-AI

**Comment by nleck** [comment]
Q2: The results are the results. If they are not acceptable we may extend the timeout and if that doesn't give what we want we will look at the NEAT-AI options.

**Comment by nleck** [comment]
Q3 Lets try upto 10 minutes if that doesn't give us a reasonable result let's think again. Not really a problem with the examples itself

**Comment by nleck** [comment]
Q4: Full sample. The Rust scorer is really quite fast on large data. 

What is the normal/expected training time for a NEAT implementation from scratch ( without cheating) ?

**Comment by stservice** [comment]
Claimed by `stservice-1778398303179`

**Comment by stservice** [comment]
## Grill-Me Round 2

**TL;DR:** Three small follow-ups: telemetry plumbing scope, README rewrite depth, and whether one 10-min run is enough evidence. Plus a quick answer to your training-time question.

### Understanding
Locked in from Round 1: delete `runMinimalSeedEvolveDir`, call `evolveDir` once with only `{ targetError: 0.001, timeoutMinutes: 10 }`, use full 28×28 / 60k samples, and quote whatever accuracy the honest run produces. Telemetry should ride on NEAT-AI library options — file follow-up issues there if hooks are missing rather than re-introducing a wrapper loop.

### Re: expected training time
Pure topology-evolving NEAT on raw MNIST (784→10, 60k samples) is hard for the algorithm — most published NEAT-style MNIST work runs for **hours on CPU**, and the original NEAT paper never tackled MNIST. HyperNEAT (indirect encoding) reached ~92% in published work; vanilla NEAT typically lands well below MLP baselines without long runs. Realistic expectation for a 10-min CPU run from scratch: somewhere between random chance and ~50%, not 95%.

### Questions

1. Telemetry plumbing in this PR?
   - [x] Ship `evolveDir` with no per-generation chart now; file a NEAT-AI issue requesting stats hooks (mirroring NEAT-AI #2629 for evolveRL); add charts in a follow-up PR
   - [ ] Block this PR until NEAT-AI exposes per-generation stats hooks for `evolveDir`
   - [ ] Use whatever stats `evolveDir` already emits (logs/return value) and embed those, even if minimal
   - [ ] other — please describe in a reply

2. README rewrite depth?
   - [ ] Keep current narrative structure, just replace the "95% target" claims with the real measured number and runtime
   - [x] Cut the README down to a short "what evolveDir produced in 10 min from a minimal seed" report
   - [ ] Keep narrative + add a "Limitations" section explaining why pure NEAT on MNIST is slow vs. an MLP baseline
   - [x] other — please describe in a reply

3. Evidence for the headline number?
   - [x] One 10-min run is enough — quote it as the result
   - [ ] Run 3× with different seeds, quote the median
   - [ ] Run 3× with different seeds, quote min/median/max
   - [ ] other — please describe in a reply

Tick the boxes that apply (the GitHub mobile app lets you tap each one), then add any free-form notes in a reply.

**⏳ Awaiting your reply.** The Vibe Coder is waiting for your response on this issue to continue grilling.

**Comment by nleck** [comment]
R2Q2: If you have identified a limitation please raise GH issues in NEAT-AI with a call back to update the examples when done.

**Comment by stservice** [comment]
## Grill-Me — Ready for Next Phase

**TL;DR:** Replace `runMinimalSeedEvolveDir` with one `Creature.evolveDir` call using `{ targetError: 0.001, timeoutMinutes: 10 }` on full 28×28 / 60k MNIST; cut README to a short real-numbers report; file NEAT-AI follow-ups for any missing telemetry hooks.

### Agreed scope
- Seed: `new Creature(784, 10)` — no `NeatOptions` overrides, no warm start.
- Single call: `Creature.evolveDir(dataDir, { targetError: 0.001, timeoutMinutes: 10 })`. Delete the chunking wrapper and its `populationSize: 12` / `maxIterations: 200` overrides.
- Full 60k train / 10k test via the existing Rust scorer.
- One 10-min run is enough — quote whatever accuracy that produces.
- README cut to a short "what evolveDir produced in 10 min from a minimal seed" report. No "95% target" framing.
- Per-generation telemetry rides on NEAT-AI library options. If hooks are missing, raise a NEAT-AI issue (mirroring NEAT-AI #2629 for evolveRL) with a call-back to update this example when they land. No wrapper loop. Charts in a follow-up PR.
- Any other limitation found during this work → raise a NEAT-AI issue with call-back, do not work around in examples.

### Assumptions
- A from-scratch NEAT run on 784→10 with 60k samples in 10 min on CPU is unlikely to reach MLP-baseline accuracy. Headline number may be low — that is accepted.

### Related open issues
- #262, #263, #264 — adaptive_mutation rewrite; share the same minimal-seed + single-`evolveDir`-call shape and "real measured numbers" README convention.
- #220, #223 — cart_pole / maze_navigation audits; same spirit (real numbers only), different paradigm (event-driven vs supervised-batch).
- #235 — supervised-batch vs event-driven docs. This PR is a canonical supervised-batch example.
- `stSoftwareAU/NEAT-AI` — no existing `evolveDir` per-generation stats issue; a follow-up will be filed with a call-back to update this example.

### How would you like to proceed?
- [ ] Apply the planning label
- [ ] Apply the work-on label

Tick the box for your choice, then open the issue on GitHub mobile, tap the labels icon, and toggle either `planning` or `work-on` to actually apply it.

**Comment by stservice** [comment]
Claimed by `stservice-1778402484102`

**Comment by stservice** [comment]
## Planning complete

Broken into four sub-issues. Two distinct asks (core code change, README rewrite), one upstream-tracking ask (file NEAT-AI follow-up), and one deferred follow-up ask (charts after upstream lands).

### Sub-issues created

| #     | Title                                                                                          | Depends on    |
| ----- | ---------------------------------------------------------------------------------------------- | ------------- |
| #270  | MNIST: direct evolveDir on full 28×28 / 60k with targetError 0.001 + timeoutMinutes 10          | —             |
| #271  | MNIST: cut README to a short measured-results report (no 95% target framing)                    | #270          |
| #272  | File NEAT-AI upstream issue: per-generation telemetry hooks for evolveDir                       | —             |
| #273  | MNIST: restore per-generation telemetry once NEAT-AI evolveDir hooks land                       | #270, #272    |

### Suggested implementation order

1. **#270 + #272 in parallel** — independent. #270 is the foundation code change; #272 just files an upstream issue at `stSoftwareAU/NEAT-AI`.
2. **#271** — once #270 has produced measured numbers from the 10-minute run, rewrite the README around them.
3. **#273** — deferred. Wait until the upstream NEAT-AI issue from #272 lands and ships in a release; then restore per-generation telemetry on top of #270's single-call shape.

```mermaid
flowchart LR
    A270["#270 Core fix<br/>(direct evolveDir, 784/60k)"] --> B271["#271 README rewrite"]
    A270 --> D273["#273 Charts<br/>(deferred follow-up)"]
    C272["#272 File NEAT-AI upstream"] --> D273
```

### Decisions and assumptions

- **Legacy modes deleted.** `MNIST_NEAT_EVOLUTION` (the in-process `evolveClassifier` mutation loop) and `MNIST_MLP_BASELINE` (`evolveMLPClassifier` + `gradient.ts`) are removed in #270. They were workarounds for the dropped "95% accuracy target" framing and assume `FEATURE_COUNT === 196`, which conflicts with the audit's `new Creature(784, 10)` seed. Cleanest path is to delete them and their tests in the same PR.
- **Down-sampling deleted.** `data.ts`'s 14×14 mean-pool is removed; `FEATURE_COUNT` becomes 784. The raw `pixels` field already carries the 28×28 source, so the SVG renderer is unaffected.
- **`evolveDir` options are literally `{ targetError: 0.001, timeoutMinutes: 10 }`.** No `populationSize`, `maxIterations`, `mutationRate`, `mutationAmount`, `costOfGrowth`, `seed`, `verbose`, `log`, `threads`, or `onTrainingEvent` keys. The audit's "no overrides" rule is enforced by an acceptance criterion.
- **Per-generation telemetry deferred.** Without an upstream `onTrainingEvent` hook for `evolveDir`, hand-rolling per-generation collection would re-introduce the chunking wrapper this issue exists to delete. Charts therefore wait for #272 to land upstream and #273 to wire it in.
- **README quotes whatever the run produces.** A from-scratch NEAT topology evolution on 784→10 with 60k samples in 10 minutes on CPU is unlikely to break chance accuracy — the parent issue's "results are the results" rule applies. #271 quotes the honest measured number, whatever it turns out to be.
- **Honesty test handling.** `readme_screenshot_honesty_test.ts` currently asserts on `fitness.svg`, `topology.svg`, `evolution.csv`, and the `#210` audit reference. #271 replaces it with assertions appropriate to the cut-down README (prediction grid + measured accuracy + #268 reference). #273 restores the telemetry assertions when the artefacts return.

This planning issue will be automatically closed by the worker once planning is complete. **Reopen this issue** to request changes to the plan.

**Comment by stservice** [comment]
## Planning Complete

Planning complete. **4 sub-issue(s)** created:

- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/273
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/272
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/271
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/270

---
🤖 Processed by: stservice

**Comment by stservice** [comment]
Planning complete — sub-issues created.
